### ISSUE TITLE
Mypy not detecting abstract methods returning undefined types

### ISSUE THREAD

**Issue by brunokc** [python/mypy#14545]
**Bug Report**

I have a class with a couple of abstract methods that return a particular custom type. That custom type is then defined right after the class. Then a global variable referencing that class is created:

**To Reproduce**

```python
# test.py
from abc import ABC, abstractmethod
from typing import NamedTuple, List

class F(ABC):
    @abstractmethod
    async def m1(self) -> Result:
        pass
    
    @abstractmethod
    async def m2(self) -> Result:
        pass

class Result(NamedTuple):
    f1: str
    f2: int
    
foo: List[F] = []
```

```
> mypy --strict test.py
Success: no issues found in 1 source file
```

```
> python test.py
Traceback (most recent call last):
  File "test.py", line 4, in <module>
    class F(ABC):
  File "test.py", line 6, in F
    async def m1(self) -> Result:
NameError: name 'Result' is not defined
```

**Expected Behavior**

I would expect mypy to detect that `Result` is not defined before its use in class `F`.

**Actual Behavior**

Mypy is happy with the code as it is, even under `--strict`, while Python can't even run the code as is. Moving the declaration of `Result` above class `F` solves the problem and both Python and mypy are then happy. 

**Your Environment**

- Mypy version used: 0.991
- Mypy command-line flags: --strict
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.9.13

**Comment by hauntsaninja** [comment]
cc @ilinum I'd expect `--enable-error-code used-before-def` to catch this on master as long as `from __future__ import annotations` doesn't apply, but it doesn't.
